### DISCUSSION TITLE
Animated images

### DISCUSSION THREAD

**Discussion by maximkrouk** [kean/NukeUI#6] [Ideas]
I created AVPlayer notes, might be helpful for playing videos:
https://www.craft.do/s/2Ic8O8TrFiK8L2

Also, AVPlayer does not fit TikTok-like dynamic previews in collectionViews, due to system limitations.

Also I guess you know about [YYImage](https://github.com/ibireme/YYImage), it plays animated images without AVPlayer, but does not support SwiftPM, maybe you can find a way to implement animated images without AVPlayer, it would be very helpful 👉👈.

**Comment by kean** [comment]
Thanks, let me go through those.

Yeah, I have a note about the limitations in the README:

> WARNING: The number of players you can have at once on-screen is limited. The limit is not documented and depends on the platform. In general, expect to have about four players playing at once.

It's a shame, but there are still plenty of use-cases where the video is useful.

> Also I guess you know about YYImage, it plays animated images without AVPlayer

Do you mean GIF? NukeUI supports them as well thanks to the built-in Gifu integration. I recently started contributing to Gifu too, so I hope more improvements are coming.

**Reply by maximkrouk** [reply]
I meant webp/apng, AFAIU NukeUI does not support it yet, but probably will in some future.

Video is definitely useful in some cases, but I just pointed some issues, that may be descussed.

Also, an issue with player limits was mentioned in [_Pinterest Medium_](https://medium.com/pinterest-engineering/building-native-video-pins-7ff89ad3ec33) →  [**Stackoverflow**](https://stackoverflow.com/questions/8608570/avplayeritem-fails-with-avstatusfailed-and-error-code-cannot-decode/9933853#9933853) → [_Apple docs_](https://developer.apple.com/library/archive/documentation/AudioVideo/Conceptual/AVFoundationPG/Articles/02_Playback.html) that
> It is not a limit on the number of instances of AVPlayer, or AVPlayerItem. Rather, it is the association of AVPlayerItem with an AVPlayer which creates a "render pipeline", and you are limited to 4 of these.
----
> So AFAIU even if 3 players are paused and disconnected from AVPlayerLayers, but 2 are playing you may get `cannot decode` error
>  _But stackoverflow may be not the best source of truth_ 🌚

**Reply by kean** [reply]
Oh yeah, animated WebP and APNG are on the list. I haven't gotten it to it yet because I personally don't use those. 

For APNG, it should be as easy as integrating [APNGKit](https://github.com/onevcat/APNGKit), which I'll most likely do soon.

**Reply by maximkrouk** [reply]
Yep, I use APNGKit and it works fine, but APNGs are much heavier than WebP 🥲

**Reply by kean** [reply]
I'm adding an API to extend `LazyImage(View)` rendering in the next release:

```swift
import SwiftSVG

// Affects both all `LazyImage` and `LazyImageView` instances
LazyImageView.registerContentView {
    if $0.type == .svg, let data = $0.data {
        UIView(SVGData: data)
    }
    return nil
}
```

**Reply by maximkrouk** [reply]
If I implement animated webp using https://github.com/SDWebImage/libwebp-Xcode, should I integrate it here, or in some kind of a plug-in for Nuke? 

> It may require implementing `AnimatedImageView`, draft implementation can be found [here](https://github.com/tattn/AnimatedWebP/blob/master/AnimatedWebP/AnimatedImageView.swift)).
> `UIImageView` has startAnimating/stopAnimating methods, but `NSImageView` doesn't so maybe you have a vision on how to integrate it to existing codebases, considered options:
> - A separate class (PlatformImageView subclass or just a view), like LazyImageView (may be not the best solution for a plugin-framework)
> - If it becomes a part of NukeUI, this new feature (LazyImageView(View)) may be useful
> - Creating a shared animation manager where you have to register ImageView and animated image (enables user to use plain ImageViews, but completely inelegant)

**Reply by kean** [reply]
Nuke has a plugin system that you can extend without changing the source. There are two steps:

**Step 1**. Register a custom decoder for webp.

There are a [couple of ways](https://kean.blog/nuke/guides/image-formats#registering-decoders) to do that. For example:

```swift
extension ImageType {
    static let webp: ImageType = "public.webp"
}

ImageDecoderRegistry.shared.register { context in
    let isWebP = context.urlResponse?.url?.absoluteString.hasSuffix("webp") ?? false
    return isWebP ? ImageDecoders.Empty(type: .webp) : nil
}
```

> Instead of using `hasSuffix`, I would suggest checking magic numbers. That's the most reliable way to determine an image format. See https://github.com/ryokosuge/Nuke-WebP-Plugin/blob/master/Source/Extensions/Data%2BWebPFormat.swift

`ImageDecoders.Empty` doesn't decode the image, but attached data to the image container (`ImageContainer`).

**Step 2**. Register your rendering system to run animations.

```swift
LazyImageView.registerContentView { container: ImageContainer in
    if container.type == .webp, let data = container.data {
        YourWebPAnimatedImageView(data: data)
    }
    return nil // Default rendering will be used
}
```

**Reply by kean** [reply]
This API `LazyImageView.registerContentView` will be available in the release I'm cutting later today.

**Reply by maximkrouk** [reply]
My decoding didn't work properly for images with alpha, so I decided to use SDWebImage for animated webp and Nuke for other cases for now ☹️

<video src="https://user-images.githubusercontent.com/40476363/122813615-cd5c9680-d2db-11eb-9c53-9fd86e1c7fab.MP4"></video>

I don't like SDWebImage APIs, but it worked, so I don't have to spend time on manual implementation yet, hopefully someday or someone else... 🙂
